File: backend/core/schemas.py
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class RateRequest:
    carrier: str
    rate_type: str
    service: str

    direction: str

    origin_country: str
    destination_country: str

    weight_kg: float

    dimensions_cm: Optional[tuple[float, float, float]] = None

    packages: int = 1

    postal_code_origin: Optional[str] = None
    postal_code_destination: Optional[str] = None

    discount_pct: Optional[float] = None

    extra: dict = field(default_factory=dict)


@dataclass
class RateResult:
    carrier: str
    rate_type: str
    service: str

    zone: str

    base_price: float

    surcharges: dict[str, float] = field(default_factory=dict)

    discount: float = 0

    total: float = 0

    currency: str = "IDR"

    notes: list[str] = field(default_factory=list)

    extra: dict = field(default_factory=dict)

File: src/core/schemas.js
/**
 * @typedef {Object} PackageItem
 * @property {number} [qty=1]
 * @property {number} weight_kg
 * @property {number} [length_cm=0]
 * @property {number} [width_cm=0]
 * @property {number} [height_cm=0]
 * @property {string} [packing_type="box"]
 * @property {string} [label=""]
 */

/**
 * @typedef {Object} RateRequest
 * @property {string} carrier - "fedex" | "ups"
 * @property {string} rate_type - "publish" | "commercial" | "a26" | "b26"
 * @property {string} service - e.g., "IP", "IE", "saver"
 * @property {string} direction - "export" | "import"
 * @property {string} origin_country
 * @property {string} destination_country
 * @property {number} weight_kg
 * @property {number[]} [dimensions_cm] - [L, W, H]
 * @property {string} [postal_code_origin]
 * @property {string} [postal_code_destination]
 * @property {number} [discount_pct=0]
 * @property {Object} extra - Extra parameters (fsi_pct, optional surcharges, packages array, etc.)
 */

/**
 * @typedef {Object} RateResult
 * @property {string} carrier
 * @property {string} rate_type
 * @property {string} service
 * @property {string} zone
 * @property {number} base_price
 * @property {Object.<string, number>} surcharges
 * @property {number} discount
 * @property {number} total
 * @property {string} currency
 * @property {string[]} notes
 * @property {Object} extra
 */

export {};
